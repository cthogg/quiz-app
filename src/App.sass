@charset "utf-8";

// Import a Google Font
@import url('https://fonts.googleapis.com/css?family=Sniglet&display=swap');
// Set your brand colors
$orange: #FFE0B3;
$light-orange: #FFEDD3;
$dark-blue: #212A51
// Update Bulma's global variables
$section-padding: 0.5rem 1.5rem;
$family-sans-serif: 'Sniglet', cursive;
$text: $dark-blue;
$widescreen-enabled: false;
$fullhd-enabled: false;
$column-gap: 1rem

// Update some of Bulma's component variables
$body-background-color: $orange;
$card-background-color: $light-orange;
$control-border-width: 2px;
$input-border-color: transparent;
$input-shadow: none;

.questionCard
    padding: 0.5rem

// Import only what you need from Bulma
@import '../node_modules/bulma/sass/utilities/_all.sass';
@import '../node_modules/bulma/sass/grid/_all.sass';
@import '../node_modules/bulma/sass/base/_all.sass';
@import '../node_modules/bulma/sass/layout/section.sass';
@import '../node_modules/bulma/sass/components/card.sass';
