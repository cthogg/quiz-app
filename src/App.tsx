import React, { useState, useEffect } from 'react'; import './App.css';
import axios from 'axios';
var he = require('he');

interface Category {
  id: number,
  name: string
}

interface Question {
  "category": string,
  "type": string,
  "difficulty": string,
  "question": string
  "correct_answer": string
  "incorrect_answers": Array<string>
}

interface Questions {
  "response_code": number,
  "results": Array<Question>
}


interface Categories {
  "trivia_categories": Array<Category>
}


export const CategoryList: React.FC<{ categories: Categories, change: Function }> = ({ categories, change }) => {
  const triviaCategories = categories.trivia_categories
  const listItems = triviaCategories.map((category: Category) =>
    <option key={category.id} value={category.id}>{category.name}</option>
  );
  return (
    <React.Fragment>
      <label htmlFor="pet-select">Choose a Category</label>
      <select id="pet-select" onChange={e => change(e.target.value)} >
        {listItems}
      </select>
    </React.Fragment>
  );
}

export const Question: React.FC<{ question: Question }> = ({ question }) => {
  return (
    <React.Fragment>
      {question.correct_answer !== undefined &&
        <React.Fragment>

          <p> {he.decode(question.question)} </p>
          <ul>
            <li>  <em> {question.correct_answer} </em> </li>
            {question.incorrect_answers.map((question, index) =>
              <li key={index}> {question} </li>
            )}
          </ul>
        </React.Fragment>

      }

    </React.Fragment>
  );
}

const QuestionsTable: React.FC<{ questions: Questions, onRowClick: Function }> = ({ questions, onRowClick }) => {
  return (
    <React.Fragment>
      <table>
        <thead>
          <tr>
            <th> Question</th>
            <th> Category</th>
            <th> Difficulty</th>

          </tr>
        </thead>
        <tbody>
          {questions.results &&
            questions.results.map((question, index) =>
              <tr key={index} onClick={() => onRowClick(question)}>
                <td>{he.decode(question.question)}</td>
                <td>{question.category}</td>
                <td>{question.difficulty}</td>
              </tr>
            )}

        </tbody>
      </table>

    </React.Fragment>
  );
}

const App: React.FC = () => {
  const emptyCategoryArray = {} as Categories
  const initialQuestions = {} as Questions
  const initialQuestion = {} as Question
  const [categories, setCategories] = useState(emptyCategoryArray);
  const [selectedCategoryId, setSelectedCategoryId] = useState(-1);
  const [questions, setQuestions] = useState(initialQuestions);
  const [selectedQuestion, setSelectedQuestion] = useState(initialQuestion);

  useEffect(() => {
    const fetchCategories = async () => {
      const result = await axios(
        `https://opentdb.com/api_category.php`,
      );
      setCategories(result.data);
    };
    fetchCategories();
  }, []);

  useEffect(() => {
    if (selectedCategoryId === -1) {
      const fetchData = async () => {
        const result = await axios(
          `https://opentdb.com/api.php?amount=10`,
        );
        setQuestions(result.data);
      };
      fetchData();

    }

    if (selectedCategoryId !== -1) {
      const fetchData = async () => {
        const result = await axios(
          `https://opentdb.com/api.php?amount=10&category=${selectedCategoryId}`,
        );
        setQuestions(result.data);
      };
      fetchData();

    }
  }, [selectedCategoryId]);

  return (
    <div className="App">
      <React.Fragment>
        {categories !== emptyCategoryArray &&
          <CategoryList categories={categories} change={setSelectedCategoryId}> </CategoryList>}
        {questions !== initialQuestions &&
          <QuestionsTable onRowClick={setSelectedQuestion} questions={questions}> </QuestionsTable>
        }
        {selectedQuestion !== initialQuestion &&
          <Question question={selectedQuestion} />

        }
      </React.Fragment>
    </div >
  );
}

export default App;
