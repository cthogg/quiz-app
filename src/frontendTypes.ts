import { APIQuestions } from "./APITypes";

export interface Answer {
  answer: string;
  isCorrect: boolean;
}

export interface Question {
  question: string;
  category: string;
  difficulty: string;
  answers: Answer[];
}

const convertBackedAnswersToFrontendAnswers = (
  backendAnswers: string[],
  isCorrect = false
): Answer[] => {
  return backendAnswers.map(answer => ({
    answer: answer,
    isCorrect: isCorrect
  }));
};

export const convertBackendQuestionsToFrontendQuestions = (
  questions: APIQuestions
): Question[] => {
  return questions.results.map(question => ({
    question: question.question,
    category: question.category,
    difficulty: question.difficulty,
    answers: convertBackedAnswersToFrontendAnswers(
      question.incorrect_answers
    ).concat(convertBackedAnswersToFrontendAnswers([question.correct_answer]))
  }));
};
