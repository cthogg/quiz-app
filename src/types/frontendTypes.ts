import { APIQuestions, APICategories } from "./APITypes";
import he from "he";

export interface Answer {
  answer: string;
  isCorrect: boolean;
}

export interface Question {
  question: string;
  category: string;
  difficulty: string;
  answers: Answer[];
}

export interface Category {
  id: number;
  name: string;
}

const convertBackedAnswersToFrontendAnswers = (
  backendAnswers: string[],
  isCorrect = false
): Answer[] => {
  return backendAnswers.map(answer => ({
    answer: he.decode(answer),
    isCorrect: isCorrect
  }));
};

export const convertBackendQuestionsToFrontendQuestions = (
  questions: APIQuestions
): Question[] => {
  return questions.results.map(question => ({
    question: he.decode(question.question),
    category: he.decode(question.category),
    difficulty: he.decode(question.difficulty),
    answers: convertBackedAnswersToFrontendAnswers(
      question.incorrect_answers
    ).concat(
      convertBackedAnswersToFrontendAnswers([question.correct_answer], true)
    )
  }));
};

export const convertBackendCategoriesToFrontendCategories = (
  categories: APICategories
): Category[] => {
  return categories.trivia_categories.map(category => ({
    ...category
  }));
};
