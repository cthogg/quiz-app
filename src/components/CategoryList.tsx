import React from "react";
import { Category } from "../frontendTypes";

const CategoryList: React.FC<{
  categories: Category[];
  change: Function;
}> = ({ categories, change }) => {
  const listItems = categories.map((category: Category) => (
    <option key={category.id} value={category.id}>
      {category.name}
    </option>
  ));
  return (
    <React.Fragment>
      <label htmlFor="pet-select">Choose a Category</label>
      <select id="pet-select" onChange={e => change(e.target.value)}>
        {listItems}
      </select>
    </React.Fragment>
  );
};

export default CategoryList;
