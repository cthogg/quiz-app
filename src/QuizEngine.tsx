import React from "react";
import CategoryList from "./components/CategoryList";
import QuestionComp from "./components/Question";
import { emptyCategoryArray } from "./App";
import { Question, Category } from "./frontendTypes";

interface QuizEngineProps {
  categories: Category[];
  questions: Question[];
  onListClick: Function;
}
const QuizEngine: React.FunctionComponent<QuizEngineProps> = ({
  categories,
  questions,
  onListClick
}: QuizEngineProps): JSX.Element => {
  return (
    <React.Fragment>
      {categories !== emptyCategoryArray && (
        <CategoryList categories={categories} change={onListClick}>
          {" "}
        </CategoryList>
      )}
      {questions.map((question, index) => (
        <QuestionComp key={index} question={question} />
      ))}
    </React.Fragment>
  );
};

export default QuizEngine;
