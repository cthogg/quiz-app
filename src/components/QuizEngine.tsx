import React from "react";
import CategoryList from "./CategoryList";
import QuestionComp from "./Question";
import { emptyCategoryArray } from "../App";
import { Question, Category } from "../types/frontendTypes";

interface QuizEngineProps {
  categories: Category[];
  questions: Question[];
  onListClick: Function;
  selectedCategoryId: number;
  isLoading: boolean;
}
const QuizEngine: React.FunctionComponent<QuizEngineProps> = ({
  categories,
  questions,
  onListClick,
  selectedCategoryId,
  isLoading
}: QuizEngineProps): JSX.Element => {
  return (
    <React.Fragment>
      <section className="section">
        <p className="is-size-2"> Quiz Pro </p>
        <p> Answer the questions below </p>
      </section>
      {categories !== emptyCategoryArray && (
        <CategoryList
          selectedCategoryId={selectedCategoryId}
          categories={categories}
          change={onListClick}
        />
      )}
      <div className="column">
        {isLoading && <section className="section"> loading</section>}
        {!isLoading && (
          <span>
            {questions.map((question, index) => (
              <QuestionComp
                key={index}
                question={question}
                questionNumber={index + 1}
              />
            ))}{" "}
          </span>
        )}
      </div>
    </React.Fragment>
  );
};

export default QuizEngine;
