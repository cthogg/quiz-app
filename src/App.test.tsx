import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import renderer from "react-test-renderer";
import CategoryList from "./components/CategoryList";
import { Question, Category } from "./frontendTypes";
import QuizEngine from "./QuizEngine";
const lodash = require.requireActual("lodash");

const cat: Category[] = [
  {
    id: 343,
    name: "bob"
  }
];
it("Category List renders correctly", () => {
  const tree = renderer
    .create(
      <CategoryList categories={cat} change={() => "hi"}>
        {" "}
      </CategoryList>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

const exampleResults: Question[] = [
  {
    category: "Entertainment: Video Games",
    difficulty: "hard",
    question:
      "In the Super Mario Bros. Series, what is Yoshi&#039;s scientific name?",
    answers: [
      { answer: "T. Yoshisaur Munchakoopas", isCorrect: true },
      { answer: "Yoshi", isCorrect: false },
      { answer: "T. Yoshisotop Munchakoopas", isCorrect: false },
      { answer: "Yossy", isCorrect: false },
      { answer: "Anne Boleyn", isCorrect: false }
    ]
  },
  {
    category: "History",
    difficulty: "medium",
    question: "Which of his six wives was Henry VIII married to the longest?",
    answers: [
      { answer: "Catherine of Aragon", isCorrect: true },
      { answer: "Jane Seymour", isCorrect: false },
      { answer: "Anne Boleyn", isCorrect: false },
      { answer: "Catherin Parr", isCorrect: false }
    ]
  }
];

it("Question Engine Renders Correctly", () => {
  lodash.shuffle = jest.fn(arr => arr);
  const tree = renderer
    .create(
      <QuizEngine
        categories={[{ id: 1, name: "sport" }]}
        onListClick={() => null}
        questions={exampleResults}
      >
        {" "}
      </QuizEngine>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly", () => {
  const tree = renderer.create(<App> </App>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});
