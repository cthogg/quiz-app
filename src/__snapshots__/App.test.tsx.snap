// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category List renders correctly 1`] = `
Array [
  <label
    htmlFor="pet-select"
  >
    Choose a Category
  </label>,
  <select
    id="pet-select"
    onChange={[Function]}
  >
    <option
      value={343}
    >
      bob
    </option>
  </select>,
]
`;

exports[`Question Engine Renders Correctly 1`] = `
Array [
  <label
    htmlFor="pet-select"
  >
    Choose a Category
  </label>,
  <select
    id="pet-select"
    onChange={[Function]}
  >
    <option
      value={1}
    >
      sport
    </option>
  </select>,
  <p>
     
    In the Super Mario Bros. Series, what is Yoshi's scientific name?
     
  </p>,
  <ul>
    <em>
       
      <li>
         
        T. Yoshisaur Munchakoopas
         
      </li>
       
    </em>
    <li>
       
      Yoshi
       
    </li>
    <li>
       
      T. Yoshisotop Munchakoopas
       
    </li>
    <li>
       
      Yossy
       
    </li>
    <li>
       
      Anne Boleyn
       
    </li>
  </ul>,
  <p>
     
    Which of his six wives was Henry VIII married to the longest?
     
  </p>,
  <ul>
    <em>
       
      <li>
         
        Catherine of Aragon
         
      </li>
       
    </em>
    <li>
       
      Jane Seymour
       
    </li>
    <li>
       
      Anne Boleyn
       
    </li>
    <li>
       
      Catherin Parr
       
    </li>
  </ul>,
]
`;

exports[`renders correctly 1`] = `
<div
  className="App"
/>
`;
